keps/1618-optional-gc-of-workloads/README.md

```diff
@@ -0,0 +1,428 @@
+decreased memory footprint for Kueue, and freeing up etcd storage. The proposed changes 
+are designed to be explicitly enabled and configured by users, ensuring backward compatibility. 
+While the primary focus is on finished Workloads, the API section can be extended to include 
+other Kueue-authored objects in the future. The implementation involves incorporating the 
+deletion logic into the reconciliation loop and includes considerations for 
+potential risks, such as handling a large number of existing finished Workloads. 
+Overall, this KEP aims to enhance Kueue's resource management and provide users with 
+more control over object retention.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently, Kueue does not delete its own Kubernetes objects, leaving it to Kubernetes' 
+garbage collector (GC) to manage the retention of objects created by Kueue, such as Pods, Jobs, and Workloads. 
+We do not set any expiration on these objects, so by default, they are kept indefinitely. Based on usage patterns, 
+including the amount of jobs created, their complexity, and frequency over time, these objects can accumulate, 
+leading to unnecessary use of etcd storage and a gradual increase in Kueue's memory footprint. Some of these objects, 
+like finished Workloads, do not contain any useful information that could be used for additional purposes, such as debugging. 
+That's why, based on user feedback in [#1618](https://github.com/kubernetes-sigs/kueue/issues/1618), a mechanism for garbage collection of finished Workloads 
+is being proposed.
+
+With this mechanism in place, deleting finished Workloads has several benefits:
+- **Deletion of superfluous information**: Finished Workloads do not store any useful runtime-related information 
+(which is stored by Jobs), nor are they useful for debugging. They could possibly be useful for auditing purposes, 
+but this is not a primary concern.
+- **Decreasing Kueue memory footprint**: Finished Workloads are loaded during Kueue's initialization and kept in memory, 
+consuming resources unnecessarily.
+- **Freeing up etcd storage**: Finished Workloads are stored in etcd memory until Kubernetes' built-in GC mechanism 
+collects them, but by default, Custom Resources are stored indefinitely.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Support the deletion of finished Workload objects.
+- Introduce configuration of global retention policies for Kueue-authored Kubernetes objects, 
+starting with finished Workloads.
```

Comment:
Does this mean that you will work on the deletion of other objects except the finished Workloads in the next iteration?

But, you mention that ''Support the deletion/expiration of any Kueue-authored Kubernetes object other than finished Workloads.'' in the Non-Goals section.

Reply:
The goal was to include certain extensibility into the design so that other objects could be added without modifying root lvl API (`Jobs`, etc.). That's the main reason why the API key for `Workloads` is nested. I haven't seen any issues for removal of completed object (other than orphaned `Workloads` #1789) but if there's any interest in extending that feature I'd be happy to work on it. I'm also happy to work on any updates to this KEP if we decide to extend the cleanup process to other objects.

Reply:
I would recommend limiting the scope for this KEP to Kueue managed objects, not only to Kueue Workload objects since actually, we aim to extend the API for the future extensibility for the other Kueue managed resources like AdmissionCheck.

The v1beta1 means that we can not bring breaking changes to APIs without bumping the API version.
So, I would like to take care of future API possibilities in this KEP.

So, my recommendation is to remove "starting with finished Workloads." from Goal, and move the statement to the proposal. And then, mention that we do not aim to add APIs for the Jobs like batch/job, RayJob as we discussed in another thread...
I'm wondering if we need to add another mechanisms for the Jobs retention in the future when we find the actual use cases.

Reply:
Got it. Let me propose another set of changes tomorrow.

Reply:
> Introduce configuration of global retention policies for Kueue-authored Kubernetes objects

This sentence matches `Workload` and `ProvisioiningRequest`. We don't need to clean up provisioning requests explicitly, because they are children of the `Workloads`.

So, for now, we only need to take care of `Workloads`. But I agree we should keep the API extensible for supporting the cleanup of finished jobs, but that can be addressed in a separate KEP.